lightning/src/debug_sync.rs

```diff
@@ -0,0 +1,216 @@
+	fn hash<H: std::hash::Hasher>(&self, hasher: &mut H) { hasher.write_u64(self.mutex_idx); }
+}
+
+pub struct Mutex<T: Sized> {
+	inner: StdMutex<T>,
+	deps: Arc<MutexMetadata>,
+	skip_checks: bool
+}
+
+#[must_use = "if unused the Mutex will immediately unlock"]
+pub struct MutexGuard<'a, T: Sized + 'a> {
+	mutex: &'a Mutex<T>,
+	lock: StdMutexGuard<'a, T>,
+}
+
+impl<'a, T: Sized> MutexGuard<'a, T> {
+	fn into_inner(self) -> StdMutexGuard<'a, T> {
+		// Somewhat unclear why this is not possible without unsafe, but oh well.
+		unsafe {
+			let v: StdMutexGuard<'a, T> = std::ptr::read(&self.lock);
+			std::mem::forget(self);
+			v
+		}
+	}
+}
+
+impl<T: Sized> Drop for MutexGuard<'_, T> {
+	fn drop(&mut self) {
+		MUTEXES_HELD.with(|held| {
+			held.borrow_mut().remove(&self.mutex.deps);
+		});
+	}
+}
+
+impl<T: Sized> Deref for MutexGuard<'_, T> {
+	type Target = T;
+
+	fn deref(&self) -> &T {
+		&self.lock.deref()
+	}
+}
+
+impl<T: Sized> DerefMut for MutexGuard<'_, T> {
+	fn deref_mut(&mut self) -> &mut T {
+		self.lock.deref_mut()
+	}
+}
+
+impl<T> Mutex<T> {
+	pub fn new(inner: T) -> Mutex<T> {
+		Mutex {
+			inner: StdMutex::new(inner),
+			deps: Arc::new(MutexMetadata {
+				locked_before: StdMutex::new(HashSet::new()),
+				mutex_idx: MUTEX_IDX.fetch_add(1, Ordering::Relaxed) as u64,
+				#[cfg(feature = "backtrace")]
+				mutex_construction_bt: Backtrace::new(),
+			}),
+			skip_checks: false,
+		}
+	}
+
+	pub fn lock<'a>(&'a self) -> LockResult<MutexGuard<'a, T>> {
+		if !self.skip_checks {
+			MUTEXES_HELD.with(|held| {
+				// For each mutex which is currently locked, check that no mutex's locked-before
+				// set includes the mutex we're about to lock, which would imply a lockorder
+				// inversion.
+				for locked in held.borrow().iter() {
+					for locked_dep in locked.locked_before.lock().unwrap().iter() {
+						if *locked_dep == self.deps {
+							#[cfg(feature = "backtrace")]
+							panic!("Tried to violate existing lockorder. Mutex which should be locked after the current lock was created at: {:?}", locked.mutex_construction_bt);
+							#[cfg(not(feature = "backtrace"))]
+							panic!("Tried to violate existing lockorder. Build with backtrace feature for more info.");
+						}
+					}
+					// Insert any already-held mutexes in our locked-before set.
+					self.deps.locked_before.lock().unwrap().insert(Arc::clone(locked));
+				}
+				held.borrow_mut().insert(Arc::clone(&self.deps));
+			});
+		}
+		self.inner.lock().map(|lock| MutexGuard { mutex: self, lock }).map_err(|_| ())
+	}
+
+	pub fn try_lock<'a>(&'a self) -> LockResult<MutexGuard<'a, T>> {
+		let res = self.inner.try_lock().map(|lock| MutexGuard { mutex: self, lock }).map_err(|_| ());
+		if !self.skip_checks && res.is_ok() {
+			MUTEXES_HELD.with(|held| {
+				// In a try-lock, only insert already-held mutexes in our locked-before set if we
+				// succeed in locking, and never check for lockinversions directly with this mutex.
```

Comment:
The first part of this comment seems obvious given this is only happens for `res.is_ok()` and the second part isn't as much. For the latter, the comment is just stating _what_ is happening and not _why_. Maybe simply say:

```rust
// A try_lock succeeding implies no lock order inversion.
```